File: HOT100/HOT100p1/9.图论/208.实现-trie-前缀树.cpp
/*
 * @lc app=leetcode.cn id=208 lang=cpp
 *
 * [208] 实现 Trie (前缀树)
 */
#include "tools.h"
// @lc code=start
class Trie {
private:
    bool isEnd;
    Trie *nxt[26];
public:
    Trie() {
        isEnd = 0;
        memset(nxt, 0, sizeof(nxt));
    }
    
    void insert(string word) {
        Trie *node = this;
        for (auto& ch : word)
        {
            if (node->nxt[ch - 'a'] == NULL) node->nxt[ch - 'a'] = new Trie();
            node = node->nxt[ch - 'a'];
        }
        node->isEnd = 1;
    }
    
    bool search(string word) {
        Trie *node = this;
        for (auto& ch : word)
        {
            node = node->nxt[ch - 'a'];
            if (node == NULL) return 0;
        }
        return node->isEnd;
    }
    
    bool startsWith(string prefix) {
        Trie *node = this;
        for (auto& ch : prefix)
        {
            node = node->nxt[ch - 'a'];
            if (node == NULL) return 0;
        }
        return 1;
    }

};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
// @lc code=end

File: HOT100/HOT100p4/208.实现-trie-前缀树.py
#
# @lc app=leetcode.cn id=208 lang=python3
#
# [208] 实现 Trie (前缀树)
#
from mytools import *
# @lc code=start
class Node:
    def __init__(self, end=False) -> None:
        self.son = [None] * 26
        self.end = end
class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        cur = self.root
        for ch in word:
            idx = ord(ch) - ord('a')
            if cur.son[idx] is None: cur.son[idx] = Node()
            cur = cur.son[idx]
        cur.end = True

    def search(self, word: str) -> bool:
        cur = self.root
        for ch in word:
            idx = ord(ch) - ord("a")
            if cur.son[idx] is None: return False
            cur = cur.son[idx]
        return cur.end
    def startsWith(self, prefix: str) -> bool:
        cur = self.root
        for ch in prefix:
            idx = ord(ch) - ord("a")
            if cur.son[idx] is None: return False
            cur = cur.son[idx]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @lc code=end

